feat: add staking rewards to vesting amounts (cip-31)

tac0turtle opened the pull request:
this pr implements cip-31

Commit:
update staking reward flow

tac0turtle committed:
add more test cases for continous

Apply suggestions from code review

Co-authored-by: Rootul P <rootulp@gmail.com>

julienrbrt committed:
chore: fix lint

julienrbrt committed:
update times

damiannolan committed:
chore: update to go1.23 in prep for celestia-core fork replace (#431)

* chore: go1.23 everywhere

* chore: use go1.23

* chore: use go-version-file in workflows

damiannolan committed:
feat: add mempool client to cometrpc interface (#435)

* feat: add mempool client to cometrpc interface

* chore: add changelog

julienrbrt committed:
feat(x/authz): add limits to grant pruning and enable message to aid manually (#442)

* feat(x/authz): add limits to grant pruning and enable message to aid manually (#18737)

remove creation of permenant and perioidic

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -247,13 +283,24 @@ func (cva ContinuousVestingAccount) GetStartTime() int64 {
+// UpdateSchedule updates the vesting schedule for a continuous vesting account.
+// It delegates to the base vesting account implementation and adjusts end time if needed.
+func (cva *ContinuousVestingAccount) UpdateSchedule(amount sdk.Coins) error {
+	// the base vesting account implementation is wrong, we should be updating the vesting schedule based on the vesting account not in the base type.
```
what does this comment mean? If the implementation of base vesting account implementation is wrong, why is the next line calling it?

tac0turtle replied:
sorry meant to remove it was a mental note.

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -247,13 +283,24 @@ func (cva ContinuousVestingAccount) GetStartTime() int64 {
 func (cva ContinuousVestingAccount) Validate() error {
-	if cva.GetStartTime() >= cva.GetEndTime() {
-		return errors.New("vesting start-time cannot be before end-time")
+	if cva.GetStartTime() > cva.GetEndTime() {
```
why was this line changed? Seems invalid to have a vesting account with start time == end time.

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -162,6 +166,38 @@ func (bva BaseVestingAccount) Validate() error {
+		return nil // No delegations, nothing to update
+	}
+
+	// Calculate what portion of the amount should be applied based on delegated vesting ratio
```
"should be applied" is vague

```suggestion
	// Calculate the amount to add to the original vesting account based on the delegated vesting ratio.
```

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -162,6 +166,38 @@ func (bva BaseVestingAccount) Validate() error {
 }
 
+// UpdateSchedule updates the vesting schedule for a base vesting account.
+// It calculates the proportion of the passed amount that should be used for updating
+// based on the ratio of delegated vesting to total delegation.
```
```suggestion
// It calculates the proportion of the passed amount that should be added to the original vesting account
// based on the ratio of delegated vesting to total delegation.
```

rootulp replied:
[quesiton] why does this care about the ratio of delegated vesting to total delegation? CIP-31 states that the lock up should be based on the amount unlocked vs locked.

Note: locked and vesting are used interchangeably but the important point is that the ratio should not care about amount delegated.

Ref: https://github.com/celestiaorg/CIPs/blob/24154ad08cdcaea005f0dee361545abd8707f3d2/cips/cip-031.md?plain=1#L70

tac0turtle replied:
the delegated free vs delegated vesting is based on the schedule. this approach uses that as a bases. we can also calculate free and not free of the orig amount. 

```
Delegation: When a user delegates tokens from a vesting account, the staking module interacts with the AccountKeeper (from x/auth). The AccountKeeper retrieves the vesting account. Methods like TrackDelegation on the vesting account (BaseVestingAccount or specific vesting types) are called. This method calculates how much of the delegation amount comes from the already vested portion (OriginalVesting minus DelegatedVesting) and how much comes from the free portion. It then updates DelegatedVesting and DelegatedFree accordingly and adjusts the overall delegated amount for the account.
```

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -162,6 +166,38 @@ func (bva BaseVestingAccount) Validate() error {
+	}
+
+	// Calculate what portion of the amount should be applied based on delegated vesting ratio
+	var updatedAmount sdk.Coins
```
[nit] this is a misleading variable name because updatedAmount isn't the new amount to use for the vesting account. Instead it is the amount to _add_ to the original vesting account.

Consider renaming `updatedAmount` => `amountToAdd`

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -387,6 +434,70 @@ func (pva PeriodicVestingAccount) Validate() error {
+		totalDelegatedForDenom := totalDelegated.AmountOf(denom)
+		delegatedVestingForDenom := pva.DelegatedVesting.AmountOf(denom)
+
+		// For new denominations, allow the amount to be free
```
what does this mean: "allow the amount to be free"?

tac0turtle replied:
for denominations received via staking that are not the staking denom would automatically be free.

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -387,6 +434,70 @@ func (pva PeriodicVestingAccount) Validate() error {
+		ratio := math.LegacyNewDecFromInt(delegatedVestingForDenom).Quo(math.LegacyNewDecFromInt(totalDelegatedForDenom))
+		amountToUse := math.LegacyNewDecFromInt(coin.Amount).Mul(ratio).RoundInt()
+
+		fmt.Println("amountToUse", amountToUse)
```
[blocking] seems residual from debugging
```suggestion
```

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -387,6 +434,70 @@ func (pva PeriodicVestingAccount) Validate() error {
+	for _, coin := range updatedAmount {
+		amtPerPeriod := coin.Amount.Quo(math.NewInt(int64(len(pva.VestingPeriods))))
+		if !amtPerPeriod.IsZero() {
+			amountPerPeriod = amountPerPeriod.Add(sdk.NewCoin(coin.Denom, amtPerPeriod))
```
[nit] it is confusing to have two variables named the same thing with one lacking some letters. Can these be disambiguated?

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -162,6 +166,38 @@ func (bva BaseVestingAccount) Validate() error {
+// UpdateSchedule updates the vesting schedule for a base vesting account.
+// It calculates the proportion of the passed amount that should be used for updating
+// based on the ratio of delegated vesting to total delegation.
+func (bva *BaseVestingAccount) UpdateSchedule(amount sdk.Coins) error {
```
[nit] why not `amount` => `rewards`

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -387,6 +434,70 @@ func (pva PeriodicVestingAccount) Validate() error {
+	for i := range pva.VestingPeriods {
+		if i == len(pva.VestingPeriods)-1 {
+			// Last period gets any remaining amount
+			remaining := updatedAmount
```
[blocking] this looks like it will be the total amount not the remaining amount.

tac0turtle replied:
this one means that the last period would get all the rewards as we dont have many periods to distribute it over.

rootulp commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -924,3 +925,585 @@ func initBaseAccount() (*authtypes.BaseAccount, sdk.Coins) {
 func TestVestingAccountTestSuite(t *testing.T) {
 	suite.Run(t, new(VestingAccountTestSuite))
 }
+
```
[blocking] why isn't there a unit test for `PeriodicVestingAccount`?

rootulp commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -924,3 +925,585 @@ func initBaseAccount() (*authtypes.BaseAccount, sdk.Coins) {
+func TestUpdateScheduleBaseVestingAcc(t *testing.T) {
+	now := tmtime.Now()
+	endTime := now.Add(24 * time.Hour)
+	bacc, initialOrigCoins := initBaseAccount() // Use a distinct name for clarity
```
Use a distinct name for what?

```suggestion
	bacc, initialOrigCoins := initBaseAccount()
```

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -518,3 +634,8 @@ func (plva PermanentLockedAccount) Validate() error {
 	return plva.BaseVestingAccount.Validate()
 }
+
+// UpdateSchedule updates the vesting schedule for a permanent locked account.
+func (plva *PermanentLockedAccount) UpdateSchedule(amount sdk.Coins) error {
+	return plva.BaseVestingAccount.UpdateSchedule(amount)
+}
```
[question] Should permanent locked accounts have CIP-31 applied to them? The rewards will never be unlocked so seems like an edge case where it doesn't make sense to lock the rewards indefinitely.

tac0turtle replied:
I implemented but would prefer to avoid it here. I can revert

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -162,6 +166,38 @@ func (bva BaseVestingAccount) Validate() error {
+	}
+
+	// Add the calculated amount to original vesting
+	bva.OriginalVesting = bva.OriginalVesting.Add(updatedAmount...)
```
[note] we're violating the vesting account specification by modifying the original vesting value.

> OV: The original vesting coin amount. It is a constant value.

Ref: https://docs.cosmos.network/main/build/modules/auth/vesting#vesting-account-specification

rootulp replied:
Does the vesting module rely on the invariant that the original vesting amount is constant?

rootulp commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -924,3 +925,585 @@ func initBaseAccount() (*authtypes.BaseAccount, sdk.Coins) {
+			rewardCoins:      sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 100)),
+			// Expecting update to proceed based on provided delegations, even if inconsistent
+			expectedVesting: initialOrigCoins.Add(sdk.NewInt64Coin(stakeDenom, 100)), // All rewards go to vesting
+			expectError:     false,                                                   // The UpdateSchedule function itself might not error here
```
It clearly doesn't error here because this test expects no error.

```suggestion
			expectError:     false,
```

rootulp commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -924,3 +925,585 @@ func initBaseAccount() (*authtypes.BaseAccount, sdk.Coins) {
+			testTime:         now.Add(12 * time.Hour).Unix(),
+		},
+		{
+			name:             "update after vesting period completed",
+			startTime:        now.Add(-48 * time.Hour).Unix(), // Start time in the past
+			endTime:          now.Add(-24 * time.Hour).Unix(), // End time in the past
+			originalVesting:  sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 100)),
+			delegatedVesting: sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 50)),
+			delegatedFree:    sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 50)),
+			rewardCoins:      sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 100)),
+			expectedVesting:  sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 150)), // Original 100 + 50 new
+			expectedEndTime:  now.Add(-24 * time.Hour).Unix(),                 // End time should remain unchanged
+			testTime:         now.Unix(),
```
If the end time is in the past, why is delegatedVesting non-zero? and why is the expecting vesting not 100?

rootulp commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -924,3 +925,585 @@ func initBaseAccount() (*authtypes.BaseAccount, sdk.Coins) {
+			delegatedVesting: sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 50)),
+			delegatedFree:    sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 50)),
+			rewardCoins:      sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 100)),
+			expectedVesting:  sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 150)), // Original 100 + 50 new
```
If the vesting has completed (endTime is in the past) then why is expectedVesting not 100?

rootulp commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -924,3 +925,585 @@ func initBaseAccount() (*authtypes.BaseAccount, sdk.Coins) {
+			expectedEndTime:  now.Unix(),
+			testTime:         now.Unix(),
+		},
+		{
+			name:             "multiple denominations in original vesting",
+			startTime:        now.Unix(),
+			endTime:          now.Add(24 * time.Hour).Unix(),
+			originalVesting:  sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 100)),
+			delegatedVesting: sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 50)),
+			delegatedFree:    sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 50)),
+			rewardCoins:      sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 100)),
+			expectedVesting:  sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 150)),
+			expectedEndTime:  now.Add(24 * time.Hour).Unix(),
+			testTime:         now.Add(12 * time.Hour).Unix(),
+		},
```
[blocking] there aren't multiple denominations here

rootulp commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -924,3 +925,585 @@ func initBaseAccount() (*authtypes.BaseAccount, sdk.Coins) {
+			delegatedVesting: sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 50)),
+			delegatedFree:    sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 50)),
+			rewardCoins:      sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 100)),
+			expectedVesting:  sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 150)), // Should still update vesting amount
```
why? If the start time == end time then the vesting is complete and I would expect expectedVesting to be 100 (the original)

rootulp commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -924,3 +925,585 @@ func initBaseAccount() (*authtypes.BaseAccount, sdk.Coins) {
+				require.Equal(t, tc.expectedVesting, cva.OriginalVesting)
+				require.Equal(t, tc.expectedEndTime, cva.EndTime)
+
+				// Check vested coins at the same future time points AFTER update
+				for i, fraction := range checkpoints {
+					checkpointTime := tc.startTime + int64(float64(duration)*fraction)
+					if checkpointTime > tc.testTime {
+						postUpdateVestedCoins := cva.GetVestedCoins(time.Unix(checkpointTime, 0))
+
+						// For cases where rewards are added (expectedVesting > originalVesting)
+						if !tc.expectedVesting.Equal(tc.originalVesting) && fraction > 0 {
+							// Verify more coins vest at each checkpoint after the update
+							for _, coin := range postUpdateVestedCoins {
+								// Find matching coin in pre-update vested coins
+								preUpdateAmt := math.ZeroInt()
+								for _, preCoin := range preUpdateVestedCoins[i] {
+									if preCoin.Denom == coin.Denom {
+										preUpdateAmt = preCoin.Amount
+										break
+									}
+								}
+
+								// If this denomination had vesting coins before the update
+								if !preUpdateAmt.IsZero() {
+									// Verify more coins vest at this checkpoint after the update
+									require.True(t, coin.Amount.GT(preUpdateAmt),
+										"Checkpoint %v (%v%%): Expected more vested coins after update. Got %v, expected more than %v",
+										checkpointTime, fraction*100, coin, preUpdateAmt)
+								}
+							}
+						}
+					}
+				}
```
What is happening here? This is so indented it's not easy to follow. Can this please be refactored?

tac0turtle replied:
this checks that the amounts unlocked at stages after an update schedule is called is greater than the original. can look at simplifying

rootulp commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -924,3 +925,585 @@ func initBaseAccount() (*authtypes.BaseAccount, sdk.Coins) {
+			delegatedVesting: sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 50)),
+			delegatedFree:    sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 50)),
+			rewardCoins:      rewardCoins,
+			expectedVesting:  initialOrigCoins.Add(sdk.NewInt64Coin(stakeDenom, 50)),
```
I would expect this to not change because the vesting time completed.

rootulp commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -924,3 +925,585 @@ func initBaseAccount() (*authtypes.BaseAccount, sdk.Coins) {
+	}
+
+	for _, tc := range testCases {
+		tc := tc // capture range variable
```
this is no longer necessary in recent Go versions

```suggestion
```

rootulp commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -924,3 +925,585 @@ func initBaseAccount() (*authtypes.BaseAccount, sdk.Coins) {
+	}
+
+	for _, tc := range testCases {
+		tc := tc // capture range variable
```
```suggestion
```

rootulp reviewed (commented):
[blocking] 
1. this needs a unit test for periodic vesting accounts
2. Why does the implementation care about delegated free vs delegated vesting? CIP-31 claims that the amount locked up should be based on amount free vs vesting regardless of whether it's delegated or not.

Ref: https://github.com/celestiaorg/CIPs/blob/24154ad08cdcaea005f0dee361545abd8707f3d2/cips/cip-031.md?plain=1#L70

damiannolan committed:
feat: add timeout info to initChain and finalizeBlock resps (#432)

* chore: go1.23 everywhere

* chore: use go1.23

* feat: add timeout info to initChain and finalizeBlock resps

* chore: use latest celestia-core branch

* chore: set prove flag to false for query tx

* chore: add changelogs

* chore: pin to HEAD of v0.38.x-celestia (core)

adjust

rootulp commented:
Synced in-person. Takeaways:
1. Remove diff from permanent and periodic vesting because they're not used on Mainnet for lockup accounts. Potentially also modify the CIP to clarify that the change only does not apply to permanent or periodic vesting accounts.
2. Instead of using the delegated free and delegated vesting, modify to use the total vesting amount and vested amount. This way the ratio is based on the total amount that has vested rather than just the staked amount that has vested.
3. Move logic to update OriginalVesting amount from the base vesting account `UpdateSchedule` to continuous and delayed vesting `UpdateSchedule` methods.

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -167,34 +167,33 @@ func (bva BaseVestingAccount) Validate() error {
+// it adds the portion of the input amount corresponding to existing denominations
+// in the original vesting schedule. If the block time is at or after the end time,
+// it does nothing.
+func (bva *BaseVestingAccount) UpdateSchedule(blockTime time.Time, amount sdk.Coins) error {
```
[nit] rename `amount` to something more descriptive. One idea `coinsToAddToOriginalVesting`

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -167,34 +167,33 @@ func (bva BaseVestingAccount) Validate() error {
-	totalDelegated := bva.DelegatedFree.Add(bva.DelegatedVesting...)
-	if totalDelegated.IsZero() {
-		return nil // No delegations, nothing to update
+// It behaves like a delayed vesting account: if the block time is before the end time,
```
How is this like a delayed vesting account? It's applicable to continuous vesting and delayed vesting.

tac0turtle replied:
removed update on base as its not needed. there is no way to track unvested vs vested in base outside of delegated free and delegated vesting. Also users will not be creating base vesting accounts

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -284,17 +283,52 @@ func (cva ContinuousVestingAccount) GetStartTime() int64 {
+		// Scale the input amount by the vesting fraction
+		addAmt := math.LegacyNewDecFromInt(coin.Amount).Mul(vestingRatio).RoundInt()
+
+		if !addAmt.IsZero() {
+			amountToAdd = amountToAdd.Add(sdk.NewCoin(coin.Denom, addAmt))
+		}
```
This conditional seems unnecessary

```suggestion
		amountToAdd = amountToAdd.Add(sdk.NewCoin(coin.Denom, addAmt))
```

tac0turtle replied:
saves a few cpu cycles but thats about it. ill remove

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -284,17 +283,52 @@ func (cva ContinuousVestingAccount) GetStartTime() int64 {
+		}
+	}
+
+	// Add the calculated unvested portion to the original vesting amount
```
unvested is incorrect here

```suggestion
	// Add the calculated portion to the original vesting amount
```

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -284,17 +283,52 @@ func (cva ContinuousVestingAccount) GetStartTime() int64 {
+	// Add the calculated unvested portion to the original vesting amount
+	if !amountToAdd.IsZero() {
+		cva.BaseVestingAccount.OriginalVesting = cva.BaseVestingAccount.OriginalVesting.Add(amountToAdd...)
 	}
```
Is this conditional necessary?

```suggestion
	cva.BaseVestingAccount.OriginalVesting = cva.BaseVestingAccount.OriginalVesting.Add(amountToAdd...)
```

rootulp commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -928,579 +928,389 @@ func TestVestingAccountTestSuite(t *testing.T) {
 
 func TestUpdateScheduleBaseVestingAcc(t *testing.T) {
```
[blocking] I think this could use a test case that updates once part way through the vesting period and then again partway through the vesting period. I'm concerned that since the `UpdateSchedule` method increases OriginalVesting amount and not the vested amount, the ratio will be incorrect for the second `UpdateSchedule` invocation.

tac0turtle replied:
we only update based on the ratio. if 100 is claimed and a user is 50% through their vest then we add 50% but the other 50 goes straight to the users account and is seen the same way as if someone were to send funds to the account, i.e not tracked. 

Ill add a test case for what you mention to make sure the calculations are the same

rootulp commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -928,579 +928,389 @@ func TestVestingAccountTestSuite(t *testing.T) {
-								}
-							}
+			// Verify results
+			require.Equal(t, tc.expectedVesting, cva.OriginalVesting, "OriginalVesting mismatch")
```
[optional] IMO this unit test should be checking the accounts `GetVestingCoins()` instead of the `OriginalVesting` amount. Also worth adding to test cases what the expected `GetVestedCoins()` are.

rootulp replied:
Later on in the test it just asserts that vested coins increases after more time has passed but the exact expected numbers would be helpful to prove this functions as intended.

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -290,8 +259,22 @@ func (cva ContinuousVestingAccount) Validate() error {
+//	The amount added to original vesting for d (ΔOV_d) is:
+//	  ΔOV_d = A_d * max(0, 1 - V_d(t) / OV_d)
+//
+// The new original vesting becomes OV'_d = OV_d + ΔOV_d.
```
The abbreviations OV, A, and t make this much harder to follow than the full words. Please update to use the full words.

rootulp commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -1314,3 +1222,191 @@ func TestUpdateScheduleDelayedVestingAcc(t *testing.T) {
+		expectedVestedLastUpdate = sdk.NewCoins(sdk.NewCoin(stakeDenom, vestedAmt))
+	}
+	actualVestedLastUpdate := cva.GetVestedCoins(lastUpdateTime)
+	require.Equal(t, expectedVestedLastUpdate, actualVestedLastUpdate, "Vested coins at last update time mismatch")
```
I don't trust this unit test because there could be a bug in the calculating of the expected values. Can you please update this test to include expected values (expected vesting coins, expected vested coins) for each test case (a.k.a for each time offset)?

rootulp commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -1314,3 +1222,191 @@ func TestUpdateScheduleDelayedVestingAcc(t *testing.T) {
+
+		// Verify the current OriginalVesting after each update
+		expectedCurrentVesting := initialVesting.AmountOf(stakeDenom).Add(cumulativeAddedVesting)
+		actualCurrentVesting := cva.GetOriginalVesting().AmountOf(stakeDenom)
```
Same for this test. Can the test cases please include expected values for the `GetVestingCoins()` and `GetVestedCoins()` rather than the unit test automatically calculating them. There could be a bug in the this test case that calculates the expected value.

tac0turtle replied:
added a simpler test case with hardcoded values. they are round numbers so the precision is not tested but since its using sdk. math its safe to assume its correct. I can test more granular numbers.

rootulp reviewed (commented):
The unit tests are too difficult to follow. How confident are you in this code? I'm extremely suspicious of the unit tests auto calculating expected values and would rather those be explicitly written in test cases. It would be especially helpful if the test case had an example that could be followed. Here's my attempt at describing an example:

1. Create continuous vesting account with end time 100 seconds from now with 100 coins. Assert GetVestingCoins = 100. Assert GetVestedCoins = 0.
2. Advance 5 seconds. Invoke UpdateSchedule with no rewards. Assert GetVestingCoins = 95. Assert GetVestedCoins = 5.
3. Advance 5 seconds. Stake 5 coins. Assert GetVestingCoins = 90. Assert GetVestedCoins = 10.
4. UpdateSchedule with 10 coins of rewards. Since 90% of tokens are still vesting, I think 9 coins should be added to original vesting and 1 should be added to vested. Assert GetVestingCoins = 99. Assert GetVestedCoins = 11.
5. Continue...

julienrbrt committed:
feat: implement cip-30 (#437)

* feat: implement cip-30

* indexed map

* update command to query rewards instead of validators

* no indexed map needed

* updates

* updates

* add tests

* Update x/distribution/keeper/hooks.go

Co-authored-by: Rootul P <rootulp@gmail.com>

* feedback

---------

Co-authored-by: Rootul P <rootulp@gmail.com>

add integration test case with bank

better the comments

go mod change

rootulp commented on tests/go.mod:
```diff
@@ -33,13 +33,14 @@ require (
 	pgregory.net/rapid v1.1.0
 )
 
+require cosmossdk.io/collections v0.4.0
```
[nit] this should be combined with the require block above for direct dependencies. 

[nit] Also there's an `// indirect` dependency in the require block above which seems unexpected. Can you please run `go mod tidy` in this directory to organize direct dependencies in one block and indirect dependencies in the block below?

rootulp commented on x/auth/vesting/exported/exported.go:
```diff
@@ -37,4 +37,6 @@ type VestingAccount interface {
 	GetDelegatedFree() sdk.Coins
 	GetDelegatedVesting() sdk.Coins
+
+	UpdateSchedule(blockTime time.Time, amount sdk.Coins) error
```
[nit] could use a godoc comment

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -248,12 +252,38 @@ func (cva ContinuousVestingAccount) GetStartTime() int64 {
 }
 
+// UpdateSchedule updates the vesting schedule for a continuous vesting account.
+// It adds a portion of the input amount to the OriginalVesting schedule.
+// The portion added corresponds to the fraction of the vesting schedule that has *not* yet occurred at the provided blockTime.
+// Denominations in the input amount that are not present in the OriginalVesting schedule are ignored.
+func (cva *ContinuousVestingAccount) UpdateSchedule(blockTime time.Time, amount sdk.Coins) error {
```
[nit][optional] `amount` => `rewards` to better communicate that this function should be invoked with rewards earned from staking

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -248,12 +252,38 @@ func (cva ContinuousVestingAccount) GetStartTime() int64 {
+		return fmt.Errorf("blockTime is after the vesting end time")
+	}
+
+	// add the total amount to the original vesting for the the token already in the vesting account
```
[blocking] I'm so confused. This comment doesn't describe the line following it. Can you please clarify what it is referring to?

tac0turtle replied:
the next four lines do what the comment says. 

we get the original vesting tokens, when we loop the amount we update the schedule with if its part of the original vesting schedule we add to it. ill a better comment to describe the flow

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -248,12 +252,38 @@ func (cva ContinuousVestingAccount) GetStartTime() int64 {
+	// add the total amount to the original vesting for the the token already in the vesting account
+	originalVesting := cva.BaseVestingAccount.GetOriginalVesting()
+
+	for _, coin := range amount {
+		origCoin := originalVesting.AmountOf(coin.Denom)
+		if origCoin.IsZero() {
+			// If the denomination wasn't originally vesting, skip it.
+			// Do not add new denominations to the original vesting schedule during an update.
+			continue
+		}
+
+		cva.BaseVestingAccount.OriginalVesting = cva.BaseVestingAccount.OriginalVesting.Add(coin)
+	}
```
[blocking] this doesn't take into account the amount of time elapsed by the vesting account. This should only add a portion of `amount` that corresponds to the unvested ratio.

tac0turtle replied:
we discussed in slack and you said you preferred this approach. im kind of lost on what the expectation is here anymore.

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -387,6 +417,13 @@ func (pva PeriodicVestingAccount) Validate() error {
+// It takes in the amount of coins from the rewards and updates the vesting schedule
+// based on the ratio of delegated vesting to total delegated coins.
+func (pva *PeriodicVestingAccount) UpdateSchedule(blockTime time.Time, amount sdk.Coins) error {
+	return nil
```
[blocking] why doesn't this return an error?

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -387,6 +417,13 @@ func (pva PeriodicVestingAccount) Validate() error {
 
+// UpdateSchedule updates the vesting schedule for a periodic vesting account.
+// It takes in the amount of coins from the rewards and updates the vesting schedule
+// based on the ratio of delegated vesting to total delegated coins.
```
[blocking] this comment does not describe the implementation.

> It takes in the amount of coins from the rewards

Let's change the argument name from `amount` to `rewards`

> based on the ratio of delegated vesting to total delegated coins.

It should be the ratio of unvested coins to total vesting coins.

Lastly periodic vesting account isn't supported so the comment should state that.

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -453,6 +490,35 @@ func (dva DelayedVestingAccount) Validate() error {
 
+// UpdateSchedule updates the vesting schedule for a delayed vesting account.
+// If the block time is before the end time, it adds the portion of the input amount
+// corresponding to existing denominations in the original vesting schedule.
```
> it adds the portion of the input amount
> // corresponding to existing denominations in the original vesting schedule.

Please clarify that this function adds a portion of rewards based on the ratio of unvested coins to total vesting coins to the original vesting schedule.

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -248,12 +252,38 @@ func (cva ContinuousVestingAccount) GetStartTime() int64 {
+// Denominations in the input amount that are not present in the OriginalVesting schedule are ignored.
+func (cva *ContinuousVestingAccount) UpdateSchedule(blockTime time.Time, amount sdk.Coins) error {
+	if blockTime.Unix() >= cva.EndTime {
+		return fmt.Errorf("blockTime is after the vesting end time")
```
why does this return an error? I would expect it to return `nil` because that is what the delayed vesting account type does.

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -453,6 +490,35 @@ func (dva DelayedVestingAccount) Validate() error {
+	originalVesting := dva.BaseVestingAccount.GetOriginalVesting()
+	var amountToAdd sdk.Coins
+
+	// Add the full amount for existing denominations since the account is entirely unvested.
```
we don't know if the account is entirely unvested. This function hasn't calculated how much time has elapsed on the vesting schedule.

tac0turtle replied:
correct, discussed in slack. will update the CIP

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -453,6 +490,35 @@ func (dva DelayedVestingAccount) Validate() error {
+		}
+	}
+
+	// Add the filtered amount to original vesting
```
what is filtered amount?
```suggestion
```

tac0turtle replied:
the amountToAdd. three lines above this comment it filters the tokens

rootulp commented on x/auth/vesting/types/vesting_account.go:
```diff
@@ -518,3 +584,8 @@ func (plva PermanentLockedAccount) Validate() error {
 	return plva.BaseVestingAccount.Validate()
 }
+
+// UpdateSchedule updates the vesting schedule for a permanent locked account.
+func (plva *PermanentLockedAccount) UpdateSchedule(blockTime time.Time, amount sdk.Coins) error {
+	return nil
+}
```
[blocking] this should probably return an error.

tac0turtle replied:
an error could cause a issue upstream as this should not error. If we error here then upstream needs to handle unimplemented updates schedules which will be treated the same as nil.

rootulp replied:
Got it. Then this comment should say it is a no-op.

rootulp commented on x/distribution/types/vesting.go:
```diff
@@ -0,0 +1,34 @@
+)
+
+// VestingAccount defines an interface used for account vesting.
+type VestingAccount interface {
```
[question] why is this type re-defined here? It looks an identical copy exists in `x/bank/types/vesting.go` and `x/auth/vesting/exported/exported.go`

tac0turtle replied:
to not have a direct dependency, this is the flow in the sdk

rootulp commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -924,3 +925,538 @@ func initBaseAccount() (*authtypes.BaseAccount, sdk.Coins) {
+		expectedEndTime int64     // EndTime should not change
+		expectError     bool      // Error expected from New... or UpdateSchedule
+	}{
+		{
+			name:            "update halfway through vesting period",
+			startTime:       now.Unix(),
+			endTime:         now.Add(24 * time.Hour).Unix(),
+			originalVesting: sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 1000)),
+			rewardCoins:     sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 100)),
+			testTime:        now.Add(12 * time.Hour).Unix(),
+			// Expected: 1000 (original) + 100 (full reward) = 1100
+			expectedVesting: sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 1100)),
+			expectedEndTime: now.Add(24 * time.Hour).Unix(),
+			expectError:     false,
+		},
```
[blocking] why does this test case expect 100 tokens are added to original vesting if it is halfway through the vesting period. I would expect only 50% of the rewards to be added to original vesting.

rootulp commented on x/auth/vesting/types/vesting_account_test.go:
```diff
@@ -924,3 +925,538 @@ func initBaseAccount() (*authtypes.BaseAccount, sdk.Coins) {
+			expectedEndTime: now.Add(24 * time.Hour).Unix(),
+			expectError:     false,
+		},
+		{
+			name:            "update 75% through vesting period",
+			startTime:       now.Unix(),
+			endTime:         now.Add(24 * time.Hour).Unix(),
+			originalVesting: sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 1000)),
+			rewardCoins:     sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 100)),
+			testTime:        now.Add(18 * time.Hour).Unix(),
+			// Expected: 1000 (original) + 100 (full reward) = 1100
+			expectedVesting: sdk.NewCoins(sdk.NewInt64Coin(stakeDenom, 1100)),
+			expectedEndTime: now.Add(24 * time.Hour).Unix(),
+			expectError:     false,
+		},
```
[blocking] this test case expects 100 coins added to original vesting when it should expect 75% of the tokens added to original vesting.

rootulp reviewed (changes requested):
I stopped reviewing after looking at two of the test cases which are clearly incorrect. Two test cases assert the same behavior even though they should assert different behavior. 

There are a number of blocking issues with this PR and it's not ready for review. It does not implement CIP-31.

tac0turtle commented:
> I stopped reviewing after looking at two of the test cases which are clearly incorrect. Two test cases assert the same behavior even though they should assert different behavior.
> 
> There are a number of blocking issues with this PR and it's not ready for review. It does not implement CIP-31.

discussed in slack, as I proposed and it was indicated that this approach was preferred by the reviewer I adjusted the implementation. I am discussing with other stake holders on the desired flow before more reviews continue as this is the third implementation of this CIP.

rootulp commented:
Yea very fair. Sorry my most recent review was definitely too harsh. I didn't realize that the most recent implementation intentionally locks up **100% of staking rewards** instead of a portion of staking rewards (like previous implementations).

> I am discussing with other stake holders on the desired flow before more reviews continue as this is the third implementation of this CIP.

Agreed. Can we put this PR in draft for now? I think it's blocked on getting alignment and updating CIP-31 with the desired flow. Specifically, should 100% of staking rewards be locked up or a portion of staking rewards based on how much has vested in the vesting account. Also which accounts are in scope (see https://github.com/celestiaorg/CIPs/issues/284)?

tac0turtle committed:
Merge branch 'release/v0.50.x-celestia' into marko/staking_rewards

tac0turtle commented:
will open a new branch to clean up commits.
